core/src/bundle_stage.rs

```diff
@@ -1307,13 +1346,28 @@ impl BundleStage {
             r_poh_recorder.would_be_leader(DROP_BUNDLE_SLOT_OFFSET * DEFAULT_TICKS_PER_SLOT);
         drop(r_poh_recorder);
 
+        let last_slot = bundle_stage_leader_stats.current_slot;
         bundle_stage_leader_stats.maybe_report(id, &working_bank_start);
 
         match (working_bank_start, would_be_leader_soon) {
             // leader now, insert new read bundles + as many as can read then return bank
             (Some(bank_start), _) => {
                 consensus_cache_updater.maybe_update(&bank_start.working_bank);
 
+                // If slot changed and currently leader, reserve space
+                if last_slot != bundle_stage_leader_stats.current_slot {
```

Comment:
imo it makes sense to move some of this to a special struct, will be easier to grasp

Reply:
all the pre-alloc logic

Reply:
will be easier to add decay in there too